File: DSA PRACTICE/LECTURE13/FindTargetIndices.cpp
// 2089. Find Target Indices After Sorting Array (leetcode problem)
//////////////////////////////////////////////
// OPTICAL SOLUTION:01(BINARY SEARCH)
class Solution {
public:
    vector<int> targetIndices(vector<int>& nums, int target) {
        
        sort(nums.begin(),nums.end());
        vector<int> ans;
        int start=0;
        int end=nums.size()-1;
        int mid=start+(end-start)/2;
        
        while(start<=end)
        {
            if(nums[mid]==target)
            {
                ans.push_back(mid);
                if(nums[mid]<nums[mid+1]) // goto to left side  
                {
                    end=mid-1;
                }
                else // goto right side
                {
                    start=mid+1;
                }
            }
            else if(nums[mid]>target)// goto left side
            {
                end=mid-1;
            }
            else // goto right side
            {
                start=mid+1;
            }
            
            mid=start+(end-start)/2;
        }
        sort(ans.begin(),ans.end());
        return ans;      
    }
};

//////////////////////////////////////////////
// OPTICAL SOLUTION:02 TC: O(n log n)
//2089. Find Target Indices After Sorting Array

//IN JAVA
class Solution {
    public List<Integer> targetIndices(int[] nums, int target) {
        int count = 0, lessthan = 0;
        for (int n : nums) {
            if (n == target) count++;
            if (n < target) lessthan++;
        }
        
        List<Integer> result = new ArrayList<>();
        for (int i = 0; i < count; i++) {
            result.add(lessthan++);
        }
        return result;
    }
}


//IN C++

class Solution {
public:
    vector<int> targetIndices(vector<int>& A, int target) {
        sort(begin(A), end(A));
        vector<int> ans;
        for (int i = 0; i < A.size(); ++i) 
        {
            if (A[i] == target) 
            {
                ans.push_back(i);
            }
        }
        return ans;
    }
};



/*
Example 1:
Input: nums = [1,2,5,2,3], target = 2
Output: [1,2]
Explanation: After sorting, nums is [1,2,2,3,5].
The indices where nums[i] == 2 are 1 and 2.

Example 2:
Input: nums = [1,2,5,2,3], target = 3
Output: [3]
Explanation: After sorting, nums is [1,2,2,3,5].
The index where nums[i] == 3 is 3.

Example 3:
Input: nums = [1,2,5,2,3], target = 5
Output: [4]
Explanation: After sorting, nums is [1,2,2,3,5].
The index where nums[i] == 5 is 4.

*/

File: DSA PRACTICE/LECTURE13/FindTargetIndices.java
// 2089. Find Target Indices After Sorting Array (leetcode problem)
// 1st Method [Sorting]:
// Code:

public List<Integer> targetIndices(int[] nums, int target) {        
        Arrays.sort(nums);
	    ArrayList<Integer> ans = new ArrayList<Integer>();
		 
        for(int i=0;i<nums.length;i++){
            if(nums[i] == target) ans.add(i);
        }
        return ans;  
}

/*
Complexity Analysis:
Time Complexity: O(nlogn) [As the array is being sorted]
Space Complexity: O(1) [No extra space used]

Explanation:

The array nums is sorted using Array.sort() method.
Then while traversing nums, if any element equals the target its index is added in the ArrayList ans.
List ans is then returned as answer.
*/
// 2nd Method [Without Sorting/Optimal]:
// Code:

    public List<Integer> targetIndices(int[] nums, int target) {
         int belowTCount=0, tCount=0;    
        for(int n : nums){
            if(n<target)
                belowTCount++;
            else if(n==target)
                tCount++;
        }
        List<Integer> ans = new ArrayList<>();
            for(int t=0;t<tCount;t++)
                ans.add(belowTCount++);
    
        return ans;
    }

/*    
Complexity Analysis:
Time Complexity: O(n) [As no sorting is done]
Space Complexity: O(1) [No extra space used]

Explanation:

While traversing nums, these variable are used:
belowTCount: To maintain the count of elements having lesser value than target.
tCount: To maintain the count of elements having value equal to target.

Now, as you can see in the for-loop, we are adding belowtCount's value(increased after every iteration) to our list, as it lets us ignore the indexes which have lesser value than the target and only add indexes of elements equal to target.

This is the same thing we would have done if the array was sorted.

List ans is then returned as answer.
*/

// methdo 03:
// code:
class Solution {
    public List<Integer> targetIndices(int[] nums, int target) {
        int count = 0, lessthan = 0;
        for (int n : nums) {
            if (n == target) count++;
            if (n < target) lessthan++;
        }
        
        List<Integer> result = new ArrayList<>();
        for (int i = 0; i < count; i++) {
            result.add(lessthan++);
        }
        return result;
    }
}
